luxjs/services/pagination.js

```diff
@@ -0,0 +1,96 @@
+// Lux pagination module for controlling the flow of
+// repeat requests to the API.
+// It can return all data at an end point or offer
+// the next page on request for the relevant component
+
+angular.module('lux.pagination', ['lux.services'])
+    .factory('luxPaginationFactory', ['$lux', function($lux) {
+
+        /**
+        * LuxPagination constructor requires three args
+        * @param scope - the angular $scope of component's directive
+        * @param target {object} - containing name and url, e.g.
+        * {name: "groups_url", url: "http://127.0.0.1:6050"}
+        * @param recursive {boolean}- set to true if you want to recursively
+        * request all data from the endpoint
+        */
+
+        function LuxPagination(scope, target, recursive) {
+            this.scope = scope;
+            this.target = target;
+            this.orgUrl = this.target.url;
+            this.api = $lux.api(this.target);
+
+            if (recursive === true) this.recursive = true;
+        }
+
+        LuxPagination.prototype.getData = function(params, cb) {
+            // getData runs $lux.api.get() followed by the component's
+            // callback on the returned data or error.
+            // it's up to the component to handle the error.
+
+            this.params = params ? params : null;
+            if (cb) this.cb = cb;
+
+            this.api.get(null, params).then(function(data) {
+
+                if (data.error) return cb(data);
+
+                this.cb(data);
+                this.updateUrls(data);
+
+            }.bind(this));
+
+        };
+
+        LuxPagination.prototype.updateUrls = function(data) {
+            // updateUrls creates an object containing the most
+            // recent last and next links from the API
+
+            if (data && data.data && data.data.last) {
+                this.emitEvent();
+                this.urls = {
+                    last: data.data.last,
+                    next: data.data.next ? data.data.next : false
+                };
+                // If the recursive param was set to true this will
+                // request data using the 'next' link
+                if (this.recursive) this.loadMore();
+            }
+
+        };
+
+        LuxPagination.prototype.emitEvent = function() {
+            // emit event if more data available, the component can
+            // listen for it and choose how to deal with it
+            this.scope.$emit('moreData');
+        };
+
+        LuxPagination.prototype.loadMore = function() {
+            // loadMore applies new urls from updateUrls to the
+            // target object and makes another getData request.
+
+            if (this.urls.next === false) {
+                this.target.url = this.urls.last;
+            } else if (this.urls.next) {
+                this.target.url = this.urls.next;
+            }
```

Comment:
I've looked into other methods of editing/updating the URL. We could pass an object with the `path` property as an option, but in this module it's preferable to use the `next`/`last` URLs provided by the api. Because those URL are full (not just the path), overwriting the target is easier and uses less code (e.g. we don't have to pass the full URLs for queries then add them to an options object).